common/include/pcl/common/impl/centroid.hpp

```diff
@@ -56,27 +56,30 @@ template <typename PointT, typename Scalar> inline unsigned int
 compute3DCentroid (ConstCloudIterator<PointT> &cloud_iterator,
                    Eigen::Matrix<Scalar, 4, 1> &centroid)
 {
-  // Initialize to 0
-  centroid.setZero ();
-
   unsigned cp = 0;
 
+  // Initialize to 0
+  Eigen::Matrix<Scalar, 3, 1> accu = Eigen::Matrix<Scalar, 3, 1>::Zero ();
+
   // For each point in the cloud
   // If the data is dense, we don't need to check for NaN
   while (cloud_iterator.isValid ())
   {
     // Check if the point is invalid
     if (pcl::isFinite (*cloud_iterator))
     {
-      centroid[0] += cloud_iterator->x;
-      centroid[1] += cloud_iterator->y;
-      centroid[2] += cloud_iterator->z;
+      accu += cloud_iterator->getVector3fMap ().template cast<Scalar> ();
       ++cp;
     }
     ++cloud_iterator;
   }
-  centroid /= static_cast<Scalar> (cp);
-  centroid[3] = 1;
+  if (cp)
+  {
+    accu /= static_cast<Scalar> (cp);
+    centroid.head (3) = accu;
```

Comment:
Use the template version `head<3>` instead

Reply:
I've fixed this and added tests which fail on master